### ISSUE TITLE
Add an info documentation for helm, and include it in MELPA

### ISSUE THREAD

**Issue by rnchzn** [emacs-helm/helm#2396]
I am thinking of generating a `texi` and `info` file from `org` file created with `M-x helm-documentation` .   

Looking at the `org` file, it seems it will need a good deal of copy-edits.  

Before I start making any changes, I would like to check if:

1. I am on right track i.e., Is `M-x helm-documentation` the right place to interest and 
2.  an in principle interest to accept the changes back in the upstream repo.

Btw,  I have already started the process. 

I had already identified the following 2 issues with `org`'s `texi` exporter, and had  them fixed in org repo.

1. https://lists.gnu.org/archive/html/emacs-orgmode/2021-04/msg00196.html
2. https://lists.gnu.org/archive/html/emacs-orgmode/2021-04/msg00206.html 

Here is a fork https://github.com/rnchzn/helm that I created a few minutes ago.  One can track any progress (or lack there of) there.

**Comment by thierryvolpiatto** [comment]
Ramesh Nedunchezian ***@***.***> writes:

> I am thinking of generating a texi and info file from org file created with `M-x helm-documentation' .

Thanks for working on this.

> Looking at the org file, it seems it will need a good deal of copy-edits. Before I start making these changes, I would like to check if:
>
>  1. I am on right track i.e., Is M-x helm-documentation the right
>  place to interest and

Yes, however this is a user oriented documentation, you will find org
file as well for a more devel oriented documentation on helm wiki,
unfortunately this documentation is still incomplete, but quite useful
for a start.

You can clone the wiki here https://github.com/emacs-helm/helm.wiki.git

You will find several other infos you may want to add to the manual.

>  
>  2. an in principle interest to accept the changes in the repo.

I am ok for this addition.

> Btw, I have already started the process. For now, I had identified 2 issues with org's texi exporter, and had them fixed in org repo. See
>
>   • https://lists.gnu.org/archive/html/emacs-orgmode/2021-04/msg00196.html
>   • https://lists.gnu.org/archive/html/emacs-orgmode/2021-04/msg00206.html
>
> Here is a fork https://github.com/rnchzn/helm that I created a few minutes ago. One can track any progress (or lack there of) in that link.

Thanks.

> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by rnchzn** [comment]
I have a created a PR https://github.com/emacs-helm/helm/pull/2397.  I will happy if you share your initial opinion.  I will be more happy happy if you could do a merge.

I hope to spend some more  time next weekend.

**Comment by rnchzn** [comment]
Could you  please review and merge this https://github.com/emacs-helm/helm/pull/2398

**Comment by thierryvolpiatto** [comment]
Ramesh Nedunchezian ***@***.***> writes:

> Could you please review and merge this #2398

I have merged your changes, thanks for your work, some comments:

- With helm-documentation keybindings are expanded which is an
  advantage.

- (info (org-texinfo-export-to-info)) is not working for me emacs-27.2,
  did I miss something?

- How to update manual when I make change in helm documentation (often).

- What do you plan for future when all manuals are merged, keep a .texi
  file and maintain it? We would write a rule in makefile to allow make
  info.

Thanks.

-- 
Thierry

**Comment by rnchzn** [comment]
Please review and merge https://github.com/emacs-helm/helm/pull/2399.  

Changes in `helm-manual-1.org' gives a good preview of what the end result is going to be.

> - With helm-documentation keybindings are expanded which is an advantage.

Info file now has an index for  commands, variables and keybindings.   These are much better than  org tables.



> - (info (org-texinfo-export-to-info)) is not working for me emacs-27.2, did I miss something? 
 
You need ox-texinfo from Org's git repo. 

> - How to update manual when I make change in helm documentation (often). - What do you plan for future when all manuals are merged, keep a .texi file and maintain it? We would write a rule in makefile to allow make info.

I have added as `make doc' target.   It generates info and html documentation.

**At some point in time in future, you  can make the  `texi2html'-generated HTML pages avaiable through project gh-pages or on a website  of your own.

These set of changes sets the tone for what will come up next.

**Comment by rnchzn** [comment]
Take look at html generated by texi2html and  let me know what you think.

**Comment by thierryvolpiatto** [comment]
Ramesh Nedunchezian ***@***.***> writes:

> Take look at html generated by texi2html and let me know what you think.

I looked at html and info files, very nice! You made a good work.

Nice work as well with make doc, I generated info+html like this.

Thanks.

I look actually at info files with helm-find-files and M-! info RET, do
you plan to install all these manuals?

By merging all these manuals in one manual?

For the html files don't know what to do with them, gh-pages are already
displaying the org files in html, what to you recommend to do with them?

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by rnchzn** [comment]
https://github.com/emacs-helm/helm/pull/2402

> I look actually at info files with helm-find-files and M-! info RET, do you plan to install all these manuals? 

Done.  Added new targets `make info` and `make install-info`.  I have not added these info targets to existing install targets.  (I don't want to mess up with your workspace.)


> By merging all these manuals in one manual?

Ultimately they will get in to doc/helm.org.  I have added this, and it is *comprehensive*.  If you arent' finding a mode, a keymap, a command or a custom var there, then there is some issue with the library (I have developed) to generate the info manual.   For all practical purposes, this is the reference manual.  All other fragments of info will be merged to this file.

   * doc/helm.org: Add it.  This is a comprehensive reference of *everything* in Helm library.
      It includes all the modes, keymaps, commands, and user options.

**Comment by thierryvolpiatto** [comment]
With your last PR I have an error because a path is hardcoded in helm.org:

emacs -q -batch -L . -l doc/ox-texinfo.el  --file=doc/helm-bugs.org --eval '(org-texinfo-export-to-texinfo)'
makeinfo --no-split doc/helm-bugs.texi -o doc/helm-bugs.info
emacs -q -batch -L . -l doc/ox-texinfo.el  --file=doc/helm-devel.org --eval '(org-texinfo-export-to-texinfo)'
makeinfo --no-split doc/helm-devel.texi -o doc/helm-devel.info
emacs -q -batch -L . -l doc/ox-texinfo.el  --file=doc/helm-manual-1.org --eval '(org-texinfo-export-to-texinfo)'
makeinfo --no-split doc/helm-manual-1.texi -o doc/helm-manual-1.info
doc/helm-manual-1.texi:493: warning: could not find @image file `./helm-M-x.txt' nor alternate text
emacs -q -batch -L . -l doc/ox-texinfo.el  --file=doc/helm-manual.org --eval '(org-texinfo-export-to-texinfo)'
makeinfo --no-split doc/helm-manual.texi -o doc/helm-manual.info
emacs -q -batch -L . -l doc/ox-texinfo.el  --file=doc/helm.org --eval '(org-texinfo-export-to-texinfo)'
Unable to read file "/home/rameshnedunchezian/src/helm/doc/doc-setup.org"
Unable to read file "/home/rameshnedunchezian/src/helm/doc/doc-setup.org"
Unable to read file "/home/rameshnedunchezian/src/helm/doc/doc-setup.org"
Undefined Org macro: kbd; aborting
Makefile:117: recipe for target 'doc/helm.texi' failed
make: *** [doc/helm.texi] Error 255


```diff
diff --git a/doc/helm.org b/doc/helm.org
index a1cc1939..970d7c6f 100644
--- a/doc/helm.org
+++ b/doc/helm.org
@@ -12453,7 +12453,7 @@ that are mentioned in the manual.  For a more complete list, use
 
 * Export Setup                                                          :noexport:
 
-#+setupfile: /home/rameshnedunchezian/src/helm/doc/doc-setup.org
+#+setupfile: ../doc/doc-setup.org
 #+options: H:4
 
 #+export_file_name: helm.texi

```

I used this to fix it, is it correct?

Thanks.

**Comment by rnchzn** [comment]
> With your last PR I have an error because a path is hardcoded in helm.org:

This should be fixed in https://github.com/emacs-helm/helm/pull/2403

    * doc/helm.org: Use relative path
    * doc/Makefile: Fix errors with `make  clean-install-info'

Let me know if you can install (and clear) the info files.  

(My installation is an inconsistent state because I was doing lots of experiments with getting the parameters to `install-info` right. I hope I got the details right this time)

**Comment by rnchzn** [comment]
This is what `info` entry will look like

```
Emacs Add-ons
* Helm Developer’s Guide: (helm-devel).
                                Helm Developer’s Guide.
* Helm Manual: (helm).          Helm Manual.
* Helm Wiki: (helm-manual-1).   Helm Wiki.
* M-x helm-documentation: (helm-manual).
                                M-x helm-documentation.
* Report Helm Bugs: (helm-bugs).
Report Helm Bugs.
```

This is the script for you to walk through ...


```sh
make clean-doc
make doc
sudo make install-info

ls -al /usr/share/info/helm*
cat /usr/share/info/dir | grep helm

info helm
info helm-devel

sudo make clean-install-info

ls -al /usr/share/info/helm*
cat /usr/share/info/dir | grep helm

info
```

**Comment by rnchzn** [comment]
Please review and pull https://github.com/emacs-helm/helm/pull/2404.

I have re-written the buffer pattern matching rules.

This is how it looks in the info buffer.  Rules and example for fuzzy matching with '^' prefix are not clear.  I would appreciate if you could add more examples. I have more questions ... but I will reserve those for future.

```
File: helm-manual.info,  Node: Pattern matching in helm-buffers,  Next: Tips,  Up: Helm Buffer

2.1 Pattern matching in helm-buffers
====================================

You can filter the buffer list based on one or more of the following
components:

  1. buffer’s ‘mode-name’
  2. its content
  3. its directory
  4. its name

   A buffer match pattern is a space-separated list of one or more
directives.  Each directive looks like

     <pattern-type><negate-may-be><pattern-value>

   ‘<pattern-type>’ determines how the ‘<pattern-value>’ is interpreted.
‘<pattern-type>’ can be one of the following:

‘*’
     Partially match ‘<pattern-value>’ against buffer’s ‘mode-name’
‘@’
     Partially match ‘<pattern-value>’ against buffer’s contents
‘/’
     Partially match ‘<pattern-value>’ against buffer’s directory
‘^’
     Fuzz-matching.  The description of what the pattern does is not
     clear.
‘​’
     an empty string.  Compare ‘<pattern-value>’ against buffer’s name.

   ‘<negate-may-be>’ determines the truth value of the match.  It can be
one of the following:

‘​’
     an empty string.  ‘<pattern-value>’ must partially match the
     corresponding attribute

‘!’
     Negate the sense of above match.

Examples
--------

Down below you see some common patterns, and how they are interpreted:

   • Match against buffer’s ‘mode-name’

     ‘*lisp’
          List only buffers whose ‘mode-name’ name matches ‘lisp’.

     ‘*!lisp’
          List only buffers whose ‘mode-name’ name does _not_ match
          ‘lisp’.

     ‘*!lisp,!sh,!fun’
          List only buffers whose mode name does _not_ match any of
          ‘lisp’, ‘sh’ or ‘fun’.

   • Match against buffer’s content

     ‘@foo’
          List only those buffers that contain ‘foo’

     ‘@!foo’
          List only those buffers that do _not_ contain ‘foo’.

     ‘@foo @bar’
          List only those buffers that contain both ‘foo’ _and_ ‘bar’.

     ‘@foo @!bar’
          List only those buffers that contain ‘foo’ but _not_ ‘bar’.

   • Match against buffer’s directory

     ‘/src’
          List only buffers whose directory matches ‘src’.

   • Match against buffer’s name in a fuzzy way

   • Match against a combination of above criterion.

     ‘*lisp ^helm @moc’
          List only buffers that are in ‘lisp’ mode, whose name starts
          start with ‘helm’ and contains ‘moc’.

     ‘*lisp ^helm moc’
          List only buffers that are in ‘lisp’ mode, whose name starts
          start with ‘helm’ and whose name contains ‘moc’

     ‘*!lisp !helm’
          List only buffers that are NOT in ‘lisp’ mode and whose name
          does NOT match ‘helm’.

     ‘/helm/ w3’
          List only buffers whose directory matches ‘helm’ and whose
          name matches ‘w3’.

```

**Comment by thierryvolpiatto** [comment]
Ramesh Nedunchezian ***@***.***> writes:

> Please review and pull #2404.
>
> I have re-written the buffer pattern matching rules.

Thanks.

> This is how it looks in the info buffer. Rules and example for fuzzy
> matching with '^' prefix are not clear.

Is this more clear?:

,----
| `helm-buffers-fuzzy-matching' turns on fuzzy matching on buffer
| names, but not on directory names or major modes.  A pattern
| starting with "^" disables fuzzy matching and matching is done
| litteraly IOW do not use regexps ("^" or whatever special
| regexp character) when you want to fuzzy match.
`----

This is not specific to buffers, this is true everywhere helm use fuzzy
matching (not sure for flex which is another beast).

Also each time a space is detected in pattern fuzzy (or flex) is
disabled and regexp matching is allowed in each pattern.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Ramesh Nedunchezian ***@***.***> writes:


> Rules and example for fuzzy matching with '^' prefix are not clear.

So this is wrong:

‘^’
     Fuzz-matching.  The description of what the pattern does is not
     clear.

It is NOT a rule, fuzzy-matching is always enable when
`helm-buffers-fuzzy-matching` is non nil.

A pattern like "hfls" will fuzzy match helm-files, but if you write
"^hfls" it will match as a regexp and will not match helm-files but any
candidate starting with "hfls".


> I would appreciate if you could add more examples. I have more
> questions ...  but I will reserve those for future.
>
> File: helm-manual.info,  Node: Pattern matching in helm-buffers,  Next: Tips,  Up: Helm Buffer
>
> 2.1 Pattern matching in helm-buffers
> ====================================
>
> You can filter the buffer list based on one or more of the following
> components:
>
>   1. buffer’s ‘mode-name’
>   2. its content
>   3. its directory
>   4. its name
>
>    A buffer match pattern is a space-separated list of one or more
> directives.  Each directive looks like
>
>      <pattern-type><negate-may-be><pattern-value>
>
>    ‘<pattern-type>’ determines how the ‘<pattern-value>’ is interpreted.
> ‘<pattern-type>’ can be one of the following:
>
> ‘*’
>      Partially match ‘<pattern-value>’ against buffer’s ‘mode-name’
> ‘@’
>      Partially match ‘<pattern-value>’ against buffer’s contents
> ‘/’
>      Partially match ‘<pattern-value>’ against buffer’s directory

Ok above is correct.

> ‘​’
>      an empty string.  Compare ‘<pattern-value>’ against buffer’s name.

I don't understand this.

>    ‘<negate-may-be>’ determines the truth value of the match.  It can be
> one of the following:
>
> ‘​’
>      an empty string.  ‘<pattern-value>’ must partially match the
>      corresponding attribute

This is not clear.

> ‘!’
>      Negate the sense of above match.

Note that negation "!" works with fuzzy and helm multi matching, but as
soon as you enter a space and a new pattern the fuzzy negation should
not work anymore, e.g. in buffers:

"!hfls" will match all buffers but helm-files, but "!hfls foo" will
match all buffers matching "foo" but the buffers matching regexp "!hfls"
(no more fuzzy here).



> Examples
> --------
>
> Down below you see some common patterns, and how they are interpreted:
>
>    • Match against buffer’s ‘mode-name’
>
>      ‘*lisp’
>           List only buffers whose ‘mode-name’ name matches ‘lisp’.
>
>      ‘*!lisp’
>           List only buffers whose ‘mode-name’ name does _not_ match
>           ‘lisp’.
>
>      ‘*!lisp,!sh,!fun’
>           List only buffers whose mode name does _not_ match any of
>           ‘lisp’, ‘sh’ or ‘fun’.
>
>    • Match against buffer’s content
>
>      ***@***.***’
>           List only those buffers that contain ‘foo’
>
>      ‘@!foo’
>           List only those buffers that do _not_ contain ‘foo’.
>
>      ***@***.*** @bar’
>           List only those buffers that contain both ‘foo’ _and_ ‘bar’.
>
>      ***@***.*** @!bar’
>           List only those buffers that contain ‘foo’ but _not_ ‘bar’.
>
>    • Match against buffer’s directory
>
>      ‘/src’
>           List only buffers whose directory matches ‘src’.
>
>    • Match against buffer’s name in a fuzzy way
>
>    • Match against a combination of above criterion.
>
>      ‘*lisp ^helm @moc’
>           List only buffers that are in ‘lisp’ mode, whose name starts
>           start with ‘helm’ and contains ‘moc’.
>
>      ‘*lisp ^helm moc’
>           List only buffers that are in ‘lisp’ mode, whose name starts
>           start with ‘helm’ and whose name contains ‘moc’
>
>      ‘*!lisp !helm’
>           List only buffers that are NOT in ‘lisp’ mode and whose name
>           does NOT match ‘helm’.
>
>      ‘/helm/ w3’
>           List only buffers whose directory matches ‘helm’ and whose
>           name matches ‘w3’.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


-- 
Thierry

**Comment by rnchzn** [comment]
Please review and pull(?) https://github.com/emacs-helm/helm/pull/2407

The most significant change is I have added `doc/helm-classes.org'.  I haven't done much EIEIO programming, so I am not sure what sort of information needs to be included.  For now I have dumped all classes and their slot (except for `helm-dabbrev-info`, which seems to be a struct)

Please take a quick look and see what sort of information would you need further.  The documenation is machine generated so using `cl-extra` library, so any request should be easily accomodated.

Apart from classes, what sort of API entry points should be documented.  May be  I can pull those as well.

**Comment by rnchzn** [comment]
> The most significant change is I have added doc/helm-classes.org.

Each class links to it's  parent and children.

The slots are ordered alphabetically.  So, it should be easy to locate a slot within the info node, without doing any C-s.  I have also added a custom class index, but it `ox-texinfo.el` doesn't accommodate non-standard indices.  I have made some local changes to the `ox-texinfo.el` for generating this index.  I will try to accommodate generating of indices via a `defadvice` for one of the `ox-texinfo` functions in the next round.
